kyo-core/shared/src/main/scala/kyo/Channel.scala

```diff
@@ -328,9 +330,22 @@ object Channel:
                 end poll
 
                 def drainUpTo(max: Int)(using AllowUnsafe) =
-                    val result = queue.drainUpTo(max)
-                    if result.exists(_.nonEmpty) then flush()
-                    result
+                    def loop(current: Chunk[A], i: Int): Result[Closed, Chunk[A]] =
+                        if i == 0 then Result.Success(current)
+                        else
+                            val next = queue.drainUpTo(i)
+                            next match
```

Comment:
`next.flatMap` should work with `@tailrec` as well

Reply:
When I try this:
```scala
@tailrec
def loop(current: Chunk[A], i: Int): Result[Closed, Chunk[A]] =
    if i == 0 then Result.Success(current)
    else
        queue.drainUpTo(i).flatMap { chunk =>
            if chunk.isEmpty then Result.Success(current)
            else
                flush()
                loop(current.concat(chunk), i - chunk.length)
        }
    end if
end loop
```

I get a tailrec error:

```
Cannot rewrite recursive call: it is not in tail position
```

Reply:
Thanks for checking! I saw it working in some code using `Maybe` but it seems there's something different here

Reply:
Yeah looks like `Maybe#flatMap` can call the mapping function in tail position since it just has to check whether the current `Maybe` is absent or present before running it, whereas `Result` has to run the mapping function in a try block before returning the result.